File: agcli/browser.py
from __future__ import annotations

from playwright.sync_api import sync_playwright


def browser_research(url: str, headless: bool = True, slow_mo_ms: int = 0) -> dict:
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=headless, slow_mo=slow_mo_ms)
        page = browser.new_page()
        page.goto(url, wait_until="networkidle")
        title = page.title()
        text = page.locator("body").inner_text()[:6000]
        links = page.locator("a[href]").evaluate_all("els => els.map(e => e.href)")
        browser.close()
    return {
        "url": url,
        "title": title,
        "text": text,
        "links": sorted(set(links))[:200],
    }

File: src/browser.js
import { chromium } from 'playwright';

export async function browserResearch(url, { headless = true, slowMoMs = 0 } = {}) {
  const browser = await chromium.launch({ headless, slowMo: slowMoMs });
  const page = await browser.newPage();
  await page.goto(url, { waitUntil: 'networkidle' });

  const title = await page.title();
  const text = (await page.locator('body').innerText()).slice(0, 6000);
  const links = await page.locator('a[href]').evaluateAll((els) =>
    [...new Set(els.map((e) => e.href))].slice(0, 200)
  );

  await browser.close();
  return { url, title, text, links };
}
